R/dock_from_renv.R

```diff
@@ -0,0 +1,90 @@
+#' @param out_dir The directory where to write the `Dockerfile`. When `NULL` no file is written. `NULL` by default.
+#' @param use_rspm Utilize the Public RStudio Package Manager for Linux binaries. Default `TRUE`.
+#'
+#' @return A R6 object of class `Dockerfile`.
+#' @details
+#'
+#' System requirements for packages are provided through RStudio Package Manager via the {pak} package. The install commands provided from pak are added as `RUN` directives within the `Dockerfile`.
+#'
+#' The R version is taken from the `renv.lock` file. Packages are installed using `renv::restore()` which ensures that the proper package version and source is used when installed.
+#'
+#' By default, `dock_from_renv()` will utilize the [public RStudio Package Manager](https://packagemanager.rstudio.com/) (PPM) to provide linux binaries. This dramatically improves install time. This is done when `use_rspm = TRUE`.  To do so, a new file `renv.lock.dock` will be created so as to not overwrite your existing `renv.lock` file. This new `renv.lock.dock` file is used to build the Docker image with PPM binaries.
+#'
+#' \donttest{
+#' dock <- dock_from_renv("renv.lock", distro = "xenial", out_dir = getwd())
+#' dock
+#' }
+#' @export
+dock_from_renv <- function(lockfile = "renv.lock", distro = "xenial", primary_repo = "CRAN",
+                           out_dir = NULL,
+                           use_rspm = TRUE) {
+
+  lock <- renv:::lockfile(lockfile)
+
+  # change to public package manager by default, if false use default repo
+  if (use_rspm) {
+    lock$repos(CRAN = glue::glue("https://packagemanager.rstudio.com/all/__linux__/{distro}/latest"))
+    # write to tempfile
+    lockfile <- paste0(lockfile, ".dock")
+    lock$write(lockfile)
+  }
+
+  # start the dockerfile
+  dock <- Dockerfile$new(gen_base_image(distro = distro, r_version = lock$version()))
+
+  distro_args <- switch(distro,
+                        centos7 = list(os = "centos", os_release = "7"),
+                        centos8 = list(os = "centos", os_release = "8"),
+                        xenial = list(os = "ubuntu", os_release = "16.04"),
+                        bionic = list(os = "ubuntu", os_release = "18.04"),
+                        focal = list(os = "ubuntu", os_release = "20.04")
+  )
+
+
+  pkgs <- names(lock$data()$Packages)
+
+  pkg_os <- lapply(pkgs, FUN = function(x) c(list(package = x), distro_args))
+
+  message(sprintf("Fetching system dependencies for %s package records.", length(pkgs)))
+
+  pkg_sysreqs <- lapply(pkg_os, function(x) do.call(pak::pkg_system_requirements, x))
+
+  pkg_installs <- unique(unlist(pkg_sysreqs))
+
+  if (distro %in% c("xenial" , "bionic" , "focal")) {
+    dock$RUN("apt-get update -y")
+  }
+
+  do.call(dock$RUN, list(pkg_installs))
+
+  renv_install <- glue::glue("install.packages('renv', repo = '{lock$repos()[{primary_repo}]}')")
+
+  dock$COPY(lockfile, "renv.lock")
+  dock$RUN(glue::glue('R -e "{renv_install}"'))
+  dock$RUN(r(renv::restore()))
```

Comment:
To be able to have one line per dependency so that we can use the power of the cache of Docker when something fails, that would be nice to extract the list of packages in the renv.lock and write one line per package as:  
```r
for (the_package in names(lock$data()$Packages)) {
  dock$RUN(r(renv::restore(packages = the_package)))
}
```
An option could allow to use this separation or not, as it may lead to many lines in the Dockerfile.

Reply:
This is an interesting suggestion. That is feasible, but as you note, that could lead to dockerfiles that are hundreds of lines long.

Reply:
This could be explored further at a late point. Given the way that `r()` works—objects are not evaluated before being written—I'm going to omit this for the time being.

Reply:
You could do it with {glue} like above, no ? 
```r
for (the_package in names(lock$data()$Packages)) {
  dock$RUN(glue::glue('Rscript -e "renv::restore(packages = {the_package})"'))
}
```

Reply:
Solid alternative!

Reply:
Well, finally this does not really take advantage of the cache, because we will have to read the "renv.lock" file before these lines of codes, and as soon as the renv.lock changes, all lines will be re-run.  
Then we do not save any time at rebuild.